### PR TITLE
Shifted methods to the correct subpackage

### LINKED ISSUE #2373: Some methods are in the wrong sub-package
1. `SHLDDRK_2N` and `SHLDDRK52` should be moved from OrdinaryDiffEqSSPRK.jl to OrdinaryDiffEqLowStorageRK.jl since they are not SSP
2. `KYK2014DGSSPRK_3S2` should be moved from OrdinaryDiffEqLowStorageRK.jl to OrdinaryDiffEqSSPRK.jl since the other SSP methods are there, even if they are low storage (which most of them are)
3. The docstring of `ssp_coefficient` should move from OrdinaryDiffEqLowStorageRK.jl to OrdinaryDiffEqSSPRK.jl or OrdinaryDiffEqCore.jl

### Details

https://github.com/SciML/OrdinaryDiffEq.jl/blob/dc0e1e733fdd79a6c643cbd0c4f4be03c934cb91/lib/OrdinaryDiffEqSSPRK/src/OrdinaryDiffEqSSPRK.jl#L102

`SHLDDRK_2N` and `SHLDDRK52` are not SSP methods but low-storage methods. There is also no `ssp_coefficient` defined for them, see 

https://github.com/SciML/OrdinaryDiffEq.jl/blob/dc0e1e733fdd79a6c643cbd0c4f4be03c934cb91/lib/OrdinaryDiffEqSSPRK/src/alg_utils.jl#L36-L52

https://github.com/SciML/OrdinaryDiffEq.jl/blob/dc0e1e733fdd79a6c643cbd0c4f4be03c934cb91/lib/OrdinaryDiffEqLowStorageRK/src/OrdinaryDiffEqLowStorageRK.jl#L107

`KYK2014DGSSPRK_3S2` is an SSP method, see also

https://github.com/SciML/OrdinaryDiffEq.jl/blob/dc0e1e733fdd79a6c643cbd0c4f4be03c934cb91/lib/OrdinaryDiffEqLowStorageRK/src/alg_utils.jl#L86-L100

### PR DESCRIPTION
**Pull request by ParamThakkar123**
## Checklist

- [X] Appropriate tests were added
- [X] Any code changes were done in a way that does not break public API
- [ ] All documentation related to code changes were updated
- [X] The new code follows the
  [contributor guidelines](https://github.com/SciML/.github/blob/master/CONTRIBUTING.md), in particular the [SciML Style Guide](https://github.com/SciML/SciMLStyle) and
  [COLPRAC](https://github.com/SciML/COLPRAC).
- [ ] Any new documentation only uses public API
  
## Additional context

fixes #2373

### CODE REVIEW

File: lib/OrdinaryDiffEqNonlinearSolve/Project.toml [hunk 1]
```diff
@@ -1,6 +1,15 @@
...
+<<<<<<< HEAD
+version = "1.0.0"
+
+[deps]
+OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
+OrdinaryDiffEqDifferentiation = "4302a76b-040a-498a-8c04-15b101fed76b"
+
+[compat]
+=======
 version = "1.1.0"
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqNonlinearSolve/Project.toml:13]:
```suggestion
version = "1.0.0"

[deps]
OrdinaryDiffEq = "1dea7af3-3e70-54e6-95c3-0bf5283fa5ed"
OrdinaryDiffEqDifferentiation = "4302a76b-040a-498a-8c04-15b101fed76b"

[compat]
version = "1.1.0"
```

File: lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl [hunk 2]
```diff
@@ -932,64 +932,85 @@ end
...
-function initialize!(integrator, cache::KYK2014DGSSPRK_3S2_Cache)
+function initialize!(integrator, cache::SHLDDRK52Cache)
     @unpack k, fsalfirst = cache
 
     integrator.kshortsize = 2
     resize!(integrator.k, integrator.kshortsize)
     integrator.k[1] = integrator.fsalfirst
     integrator.k[2] = integrator.fsallast
     integrator.f(integrator.fsalfirst, integrator.uprev, integrator.p, integrator.t) # FSAL for interpolation
-    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 1)
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl:970]: why is this deleted?

File: lib/OrdinaryDiffEqNonlinearSolve/Project.toml (hunk 1 above)
Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqNonlinearSolve/Project.toml]: Sorry. I overlooked the merge conflict part. Fixed that

File: lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl [hunk 3]
```diff
@@ -932,64 +932,85 @@ end
...
+    u = u + β5 * tmp
+
+    integrator.fsallast = f(u, p, t + dt) # For interpolation, then FSAL'd
     integrator.k[1] = integrator.fsalfirst
-    integrator.k[2] = f(integrator.u, p, t + dt) # For interpolation, then FSAL'd
-    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 3)
-    integrator.fsallast = integrator.k[2]
-    return nothing
+    integrator.k[2] = integrator.fsallast
+    integrator.u = u
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl:971]: why was this changed?

File: lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl [hunk 4]
```diff
@@ -932,34 +932,48 @@ end
...
+    tmp = α4 * tmp + dt * f(u, p, t + c4 * dt)
+    u = u + β4 * tmp
+    # u5 = u
+    tmp = α5 * tmp + dt * f(u, p, t + c5 * dt)
+    u = u + β5 * tmp
+
+    integrator.fsallast = f(u, p, t + dt) # For interpolation, then FSAL'd
     integrator.k[1] = integrator.fsalfirst
     integrator.k[2] = f(integrator.u, p, t + dt) # For interpolation, then FSAL'd
     OrdinaryDiffEqCore.increment_nf!(integrator.stats, 3)
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl:971]: This is not correct.

File: lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl [hunk 5]
```diff
@@ -968,28 +982,188 @@ function initialize!(integrator, cache::KYK2014DGSSPRK_3S2_Cache)
...
+    f(k, u, p, t + c5 * dt)
+    @.. thread=thread tmp=α5 * tmp + dt * k
+    @.. thread=thread u=u + β5 * tmp
     stage_limiter!(u, integrator, p, t + dt)
     step_limiter!(u, integrator, p, t + dt)
-    f(integrator.k[2], u, p, t + dt) # For interpolation, then FSAL'd
-    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 3)
-    return nothing
+
+    f(k, u, p, t + dt)
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl:1018]: the postamble is missing?

File: lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl [hunk 6]
```diff
@@ -968,28 +982,188 @@ function initialize!(integrator, cache::KYK2014DGSSPRK_3S2_Cache)
...
+        # u5 = u
+        tmp = α52 * tmp + dt * f(u, p, t + c52 * dt)
+        u = u + β52 * tmp
+        tmp = α62 * tmp + dt * f(u, p, t + c62 * dt)
+        u = u + β62 * tmp
+    end
+
+    integrator.fsallast = f(u, p, t + dt) # For interpolation, then FSAL'd
+    integrator.k[1] = integrator.fsalfirst
+    integrator.k[2] = integrator.fsallast
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowStorageRK/src/low_storage_rk_perform_step.jl:1084]: This is clearly missing the increment again

**Comment by ParamThakkar123** [comment]
But I made no change to any of these lines. Just moved them from one file to another just as they were

**Comment by ParamThakkar123** [comment]
image

@ChrisRackauckas Just a small question, every solver has a `dts` variable. I get a `dts` undefined error here. What the `dts` should be in this case ?

**Comment by ChrisRackauckas** [comment]
It should be the same dts as before it moved.

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas can you please review this if everything is done right ?

**Comment by ParamThakkar123** [comment]
8 tests have failed but they have errors in some other files and not in low_storage and ssprk files

**Comment by ChrisRackauckas** [comment]
https://github.com/SciML/OrdinaryDiffEq.jl/actions/runs/10535436217/job/29194409941?pr=2408#step:6:1074

**Comment by ChrisRackauckas** [comment]
Looks like you didn't move some of `alg_utils.jl`

**Comment by ChrisRackauckas** [comment]
```julia
@muladd function perform_step!(integrator, cache::SHLDDRK52ConstantCache,
        repeat_step = false)
    @unpack t, dt, uprev, u, f, p = integrator
    @unpack α2, α3, α4, α5, β1, β2, β3, β4, β5, c2, c3, c4, c5 = cache

    # u1
    tmp = dt * integrator.fsalfirst
    u = uprev + β1 * tmp
    # u2
    tmp = α2 * tmp + dt * f(u, p, t + c2 * dt)
    u = u + β2 * tmp
    # u3
    tmp = α3 * tmp + dt * f(u, p, t + c3 * dt)
    u = u + β3 * tmp
    # u4
    tmp = α4 * tmp + dt * f(u, p, t + c4 * dt)
    u = u + β4 * tmp
    # u5 = u
    tmp = α5 * tmp + dt * f(u, p, t + c5 * dt)
    u = u + β5 * tmp

    integrator.fsallast = f(u, p, t + dt) # For interpolation, then FSAL'd
    integrator.k[1] = integrator.fsalfirst
    integrator.k[2] = f(integrator.u, p, t + dt) # For interpolation, then FSAL'd
    OrdinaryDiffEqCore.increment_nf!(integrator.stats, 3)
    integrator.fsallast = integrator.k[2]
    return nothing
end
```

vs 

```julia
@muladd function perform_step!(integrator, cache::SHLDDRK52ConstantCache,
        repeat_step = false)
    @unpack t, dt, uprev, u, f, p = integrator
    @unpack α2, α3, α4, α5, β1, β2, β3, β4, β5, c2, c3, c4, c5 = cache

    # u1
    tmp = dt * integrator.fsalfirst
    u = uprev + β1 * tmp
    # u2
    tmp = α2 * tmp + dt * f(u, p, t + c2 * dt)
    u = u + β2 * tmp
    # u3
    tmp = α3 * tmp + dt * f(u, p, t + c3 * dt)
    u = u + β3 * tmp
    # u4
    tmp = α4 * tmp + dt * f(u, p, t + c4 * dt)
    u = u + β4 * tmp
    # u5 = u
    tmp = α5 * tmp + dt * f(u, p, t + c5 * dt)
    u = u + β5 * tmp

    integrator.fsallast = f(u, p, t + dt) # For interpolation, then FSAL'd
    integrator.k[1] = integrator.fsalfirst
    integrator.k[2] = integrator.fsallast
    integrator.u = u
end
```

They aren't the same code. `OrdinaryDiffEqCore.increment_nf!(integrator.stats, 3)` obviously is not correct because you can see in one screen that there are more then 3 fevals.

Just start from scratch. This is just copy-pasting code over to the other file. It shouldn't be changing the function definitions at all.

File: lib/OrdinaryDiffEqSSPRK/src/alg_utils.jl [hunk 7]
```diff
@@ -30,8 +30,21 @@ alg_order(alg::SSPRKMSVS43) = 3
...
+`dt` with `alg` can be written as a convex combination of explicit Euler steps
+with step sizes `cᵢ * dt`, the SSP coefficient is the minimal value of `1/cᵢ`.
+
+# Examples
+
+```julia-repl
+julia> ssp_coefficient(SSPRK104())
+6
+```
+"""
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqSSPRK/src/alg_utils.jl:47]: This needs to go onto the basic dispatch in order to work. It should be on OrdinaryDiffEqCore's

**Review by ChrisRackauckas — APPROVED** [review]

File: lib/OrdinaryDiffEqSSPRK/src/alg_utils.jl (hunk 7 above)
Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqSSPRK/src/alg_utils.jl]: Should I shift this docstring to OrdinaryDiffEqCore?

**Comment by ChrisRackauckas** [comment]
I fixed up that last bit, if tests pass then it's good to merge.

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas Can you please consider taking a look at #2407 . I am trying to fix some errors there. But just needed your help with one of the errors.

**Comment by ChrisRackauckas** [comment]
Real test failures https://github.com/SciML/OrdinaryDiffEq.jl/actions/runs/10540602273/job/29205446454?pr=2408#step:6:993

**Comment by ParamThakkar123** [comment]
@ChrisRackauckas @oscardssmith can you please tell me what should be done for the failing tests here?? Should I start from scratch??

**Comment by ParamThakkar123** [comment]
> Real test failures https://github.com/SciML/OrdinaryDiffEq.jl/actions/runs/10540602273/job/29205446454?pr=2408#step:6:993

what should be the order of this algorithm ? 3 ?

**Comment by ChrisRackauckas** [comment]
Those algorithm orders exist, since those tests passed before.. Just make sure you've moved alg_utils appropriately.

File: lib/OrdinaryDiffEqNonlinearSolve/test/newton_tests.jl [hunk 8]
```diff
@@ -1,4 +1,4 @@
-using OrdinaryDiffEqCore
+using OrdinaryDiffEq
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqNonlinearSolve/test/newton_tests.jl:1]: why was this changed?

File: lib/OrdinaryDiffEqLowStorageRK/test/ode_low_storage_rk_tests.jl [hunk 9]
```diff
@@ -41,6 +41,40 @@ test_problems_nonlinear = [prob_ode_nonlinear, prob_ode_nonlinear_inplace]
...
+    sim = test_convergence(dts, prob, alg)
+    @test_broken sim.𝒪est[:final]≈OrdinaryDiffEqSSPRK.alg_order(alg) atol=testTol
+end
+for prob in test_problems_linear
+    sim = test_convergence(dts, prob, alg)
+    @test_broken sim.𝒪est[:final]≈OrdinaryDiffEqSSPRK.alg_order(alg) atol=testTol
+end
+for prob in test_problems_nonlinear
+    sim = test_convergence(dts, prob, alg)
+    @test_broken sim.𝒪est[:final]≈OrdinaryDiffEqSSPRK.alg_order(alg) atol=testTol
```
Comment (ChrisRackauckas) [inline lib/OrdinaryDiffEqLowStorageRK/test/ode_low_storage_rk_tests.jl:75]:
```suggestion
    @test_broken sim.𝒪est[:final]≈OrdinaryDiffEqLowStorageRK.alg_order(alg) atol=testTol
end
for prob in test_problems_linear
    sim = test_convergence(dts, prob, alg)
    @test_broken sim.𝒪est[:final]≈OrdinaryDiffEqLowStorageRK.alg_order(alg) atol=testTol
end
for prob in test_problems_nonlinear
    sim = test_convergence(dts, prob, alg)
    @test_broken sim.𝒪est[:final]≈OrdinaryDiffEqLowStorageRK.alg_order(alg) atol=testTol
```

File: lib/OrdinaryDiffEqNonlinearSolve/test/newton_tests.jl (hunk 8 above)
Comment (ParamThakkar123) [reply lib/OrdinaryDiffEqNonlinearSolve/test/newton_tests.jl:1]: Not sure. I haven't made any change to it.
